src/main/java/org/apache/sysml/parser/BuiltinFunctionExpression.java

```diff
@@ -1792,6 +1801,12 @@ else if ( functionName.equals("bitwShiftR") )
 			bifop = Expression.BuiltinFunctionOp.BITWSHIFTR;
 		else if ( functionName.equals("ifelse") )
 			bifop = Expression.BuiltinFunctionOp.IFELSE;
+		else if (functionName.equals("eval")) {
+			bifop = Expression.BuiltinFunctionOp.EVAL;
+			//Need to cast the first param "name of the referenced func" into a string identifier
+			DataIdentifier func = (DataIdentifier) paramExprsPassed.get(0).getExpr();
+			paramExprsPassed.get(0).setExpr(new StringIdentifier(func.toString(), func));
```

Comment:
Please try to avoid this explicit modification. This should not be necessary and would probably crash if the argument is an expression such as "foo" + X[i,1]. Maybe you can add a test for such expressions to your unit test.

Reply:
I have a question about the first argument of eval func. Should it be better to always have a string of func name as the first argument, i.e, eval("f1", X, y)? Because if it accepts any expression as the first argument, i.e, eval("f1" + X[i, 1], X, y), it will be complicated to deal with this expression of any type. And in fact, I think will can use it in an another way, i.e, eval("f1", X, y) + X[i, 1].

Reply:
Let us separate two aspects: the type of the input, and the script level restrictions. The type should indeed always be a string which also guarantees that you will always get a string during runtime. 

However, we should not artificially restrict the possibility to have arbitrary expressions as input at script level because the common case will be either selections eval(X[i,]) or concatenations eval("foo"+i). These expressions will be handled automatically by the compiler, as done for any other builtin function - so you don't need to worry about it. 

Similarly, we will also allow eval in expression, but as mentioned before I will add a rewrite that splits the DAGs after eval in order to create a hook for dynamic recompilation because the sizes (dimensions and sparsity) of an unknown function will be unknown. Without dynamic recompilation this would lead to unnecessary distributed operations. In your tests you did not encounter this issue because toString is only supported as in-memory operation.